// @citizensadvice/design-system/lib/greedy-nav/Config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultConfig = void 0;
var defaultConfig = {
    initClass: 'js-CadsGreedyNav',
    mainNavWrapper: '.js-cads-greedy-nav',
    mainNav: 'ul',
    navDropdownClassName: 'cads-greedy-nav__dropdown',
    navDropdownToggleClassName: 'cads-greedy-nav__dropdown-toggle',
    navDropdownToggleAriaLabel: 'More navigation options',
    navDropdownLabel: 'More',
    navDropdownLabelActive: 'Close',
    navDropdownBreakpointLabel: 'menu',
    breakPoint: 0,
    throttleDelay: 50,
    /* Offset pixels add a tolerance to when an item is removed from the nav and put in the dropdown.
     * Aligning the nav with the grid in NP-1026 makes the contents of the nav 2px too wide for
     * GreedyNav's calculations and puts the last nav item in the dropdown at lg.  This offset prevents
     * that from occuring whilst maintainng the otherwise correct behaviour of GreedyNav.
     */
    offsetPixels: -10,
    count: true,
    // Callbacks
    moved: function () { return null; },
    movedBack: function () { return null; },
};
exports.defaultConfig = defaultConfig;
